### PR TITLE
Actor groups and group sync helpers

### PR DESCRIPTION
**Pull request by AndrewDryga**

### CODE REVIEW

File: elixir/apps/domain/lib/domain/actors.ex [hunk 1]
```diff
@@ -1,8 +1,93 @@
...
+      |> Authorizer.for_subject(subject)
+      |> Repo.list()
+    end
+  end
+
+  def new_group(attrs \\ %{}) do
+    change_group(%Group{}, attrs)
+  end
+
+  def upsert_provider_groups(%Auth.Provider{} = provider, attrs_by_provider_identifier) do
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors.ex:30]: This function will be finished in a separate PR.

File: elixir/apps/domain/test/support/fixtures/actors_fixtures.ex [hunk 2]
```diff
@@ -3,6 +3,66 @@ defmodule Domain.ActorsFixtures do
...
+        AuthFixtures.create_subject(identity)
+      end)
+
+    attrs = group_attrs(attrs)
+
+    {:ok, group} = Actors.create_group(attrs, subject)
+    group
+  end
+
+  # def create_provider_group(attrs \\ %{}) do
```
Comment (AndrewDryga) [inline elixir/apps/domain/test/support/fixtures/actors_fixtures.ex:48]: This will be finished in a separate PR.

**Comment by AndrewDryga** [comment]
I need to add a few tests but should be good for review.

File: elixir/apps/domain/lib/domain/jobs/executors/global.ex [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  end
+
+  @impl true
+  def init({{module, function}, interval, config}) do
+    name = global_name(module, function)
+
+    # `random_notify_name` is used to avoid name conflicts in a cluster during deployments and
+    # network splits, it randomly selects one of the duplicate pids for registration,
+    # and sends the message {global_name_conflict, Name} to the other pid so that they stop
+    # tying to claim job queue leadership.
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/jobs/executors/global.ex]:
```suggestion
    # trying to claim job queue leadership.
```

File: elixir/apps/domain/lib/domain/actors/group/changeset.ex [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  @fields ~w[name]a
+
+  def upsert_conflict_target do
+    {:unsafe_fragment,
+     "(account_id, provider_id, provider_identifier) " <>
+       "WHERE deleted_at IS NULL AND provider_id IS NOT NULL AND provider_identifier IS NOT NULL"}
+  end
+
+  # We do not update the `name` field because we allow to manually override it in the UI
+  # for usability reasons when the provider uses group names that can make people confused
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/actors/group/changeset.ex:15]: 👍

File: elixir/apps/domain/lib/domain/actors.ex [hunk 5]
```diff
@@ -1,8 +1,93 @@
...
+    |> case do
+      {:ok, %{group: group}} ->
+        {:ok, group}
+
+      {:error, _step, changeset, _effects_so_far} ->
+        {:error, changeset}
+    end
+  end
+
+  def create_group(attrs, %Auth.Subject{} = subject) do
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/actors.ex:62]: Doesn't have to be in this PR (apologies if I missed it), but just checking if we are still planning to lock actor memberships to groups of the same provider? I.e. so we avoid weird edge cases with manually-added actors to provider-synced groups.

File: elixir/apps/domain/lib/domain/auth.ex
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth.ex:1]: Nicely architected 😎

File: elixir/apps/domain/lib/domain/auth.ex [hunk 6]
```diff
@@ -261,6 +404,7 @@ defmodule Domain.Auth do
     salt = Keyword.fetch!(config, :salt)
     # TODO: we don't want client token to be invalid if you reconnect client from a different ip,
     # for the clients that move between cellular towers
+    # TODO: we want to show all sessions in a UI so persist them to DB
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth.ex:413]: 👍 Will be required for audit reports in some form, could start with a basic export to CSV function as an MVP

File: elixir/apps/domain/lib/domain/auth/adapters/google_workspace.ex [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    changeset
+    |> Domain.Changeset.cast_polymorphic_embed(:adapter_config,
+      required: true,
+      with: fn current_attrs, attrs ->
+        Ecto.embedded_load(GoogleWorkspace.Settings, current_attrs, :json)
+        |> OpenIDConnect.Settings.Changeset.changeset(attrs)
+      end
+    )
+
+    # TODO: validate received scopes and show an error if there are missing ones
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/auth/adapters/google_workspace.ex:54]: claims?

```suggestion
    # TODO: validate received claims and show an error if there are missing ones
```

File: elixir/apps/domain/lib/domain/jobs/executors/global.ex [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+  and then updates the processing flag on job is completed:
+
+      UPDATE my_table SET processed_at = NOW() WHERE id = ?;
+
+  it is also recommended to cancel jobs to prevent them from being executed indefinitely:
+
+      UPDATE my_table SET processing_cancelled_at = NOW() WHERE id = ?;
+
+  Even though this does not prove fully fledged job queue, it is a good enough solution
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/jobs/executors/global.ex]:
```suggestion
  Even though this does not provide a fully fledged job queue, it is a good enough solution
```

File: elixir/apps/domain/lib/domain/jobs/executors/global.ex [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  def init({{module, function}, interval, config}) do
+    name = global_name(module, function)
+
+    # `random_notify_name` is used to avoid name conflicts in a cluster during deployments and
+    # network splits, it randomly selects one of the duplicate pids for registration,
+    # and sends the message {global_name_conflict, Name} to the other pid so that they stop
+    # tying to claim job queue leadership.
+    with :no <- :global.register_name(name, self(), &:global.random_notify_name/3),
+         pid when is_pid(pid) <- :global.whereis_name(name) do
+      # we monitor the leader process so that we start a race to become a new leader with it's down
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/jobs/executors/global.ex]:
```suggestion
      # we monitor the leader process so that we start a race to become a new leader when it's down
```

File: elixir/apps/domain/lib/domain/jobs/executors/global.ex [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  end
+
+  def handle_info(:tick, {{{module, function}, interval, config}, :leader} = state) do
+    :ok = execute_handler(module, function, config)
+    :ok = schedule_tick(interval)
+    {:noreply, state}
+  end
+
+  # tick is scheduled by using a timeout message instead of `:timer.send_interval/2`,
+  # because we jobs to overlap if they take too long to execute
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/jobs/executors/global.ex]:
```suggestion
  # because we don't want jobs to overlap if they take too long to execute
```

File: elixir/apps/web/lib/web/components/layouts/app.html.heex
Comment (jamilbk) [inline elixir/apps/web/lib/web/components/layouts/app.html.heex:1]: Thank you for cleaning this up 😅

**Review by jamilbk — APPROVED** [review]
Just some minor typo suggestions and a couple questions, but otherwise this beast LGTM

File: .github/workflows/elixir.yml [hunk 11]
```diff
@@ -305,7 +305,7 @@ jobs:
           - 8200:8200/tcp
         options: --cap-add=IPC_LOCK
     steps:
-      - uses: nanasess/setup-chromedriver@v1
+      - uses: nanasess/setup-chromedriver@v2
```
Comment (jamilbk) [inline .github/workflows/elixir.yml:308]: Fixed in #1827 to unblock other PRs

File: elixir/apps/domain/lib/domain/actors/actor/changeset.ex [hunk 12]
```diff
@@ -1,16 +1,15 @@
...
-  def create_changeset(%Auth.Provider{} = provider, attrs) do
+  def create_changeset(account_id, attrs) do
     %Actors.Actor{}
     |> cast(attrs, ~w[type name]a)
     |> validate_required(~w[type name]a)
-    |> put_change(:account_id, provider.account_id)
+    |> put_change(:account_id, account_id)
   end
 
-  def set_actor_type(actor, type) when type in [:account_user, :account_admin_user] do
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/actors/actor/changeset.ex:13]: Just curious why these are no longer needed?

File: elixir/apps/domain/lib/domain/actors/group/changeset.ex (hunk 4 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/actors/group/changeset.ex:15]: Just wanted to make sure I'm following here....Is this saying that when we're syncing groups from a customers IDP it won't be syncing the name of the group if the admin has manually set the name in Firezone?

File: elixir/apps/domain/lib/domain/auth.ex [hunk 13]
```diff
@@ -24,6 +24,62 @@ defmodule Domain.Auth do
...
+        {:error, reason} ->
+          {:error, reason}
+      end
+    else
+      false -> {:error, :not_found}
+      other -> other
+    end
+  end
+
+  def fetch_provider_by_id(id) do
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/auth.ex:52]: I feel like I've noticed a lot of the other models use exclamation marks at the end of methods that aren't checking permissions.  Just wondering if this method could use one as well?

File: elixir/apps/domain/lib/domain/config/definitions.ex [hunk 14]
```diff
@@ -444,13 +444,13 @@ defmodule Domain.Config.Definitions do
...
-    {
-      :array,
-      ",",
-      {:parameterized, Ecto.Enum,
-       Ecto.Enum.init(values: ~w[email openid_connect userpass token]a)}
-    },
-    default: ~w[email openid_connect token]a
+    {:array, ",", {:parameterized, Ecto.Enum, Ecto.Enum.init(values: ~w[
+      email
+      openid_connect google_workspace
```
Comment (bmanifold) [inline elixir/apps/domain/lib/domain/config/definitions.ex:449]: Just curious if these two were meant to be on the same line?

File: elixir/apps/domain/lib/domain/actors.ex (hunk 5 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/actors.ex:62]: @jamilbk people should be able to edit groups managed manually, but it should not let you manually manage synced stuff.

Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/actors.ex:62]: I will add a couple of tests to cement this decision in the code.

File: elixir/apps/web/lib/web/auth.ex [hunk 15]
```diff
@@ -169,10 +207,14 @@ defmodule Web.Auth do
...
 
+  def on_mount(:mount_account, params, session, socket) do
+    {:cont, mount_account(socket, params, session)}
+  end
+
   def on_mount(:ensure_authenticated, params, session, socket) do
     socket = mount_subject(socket, params, session)
 
-    if socket.assigns.subject do
+    if socket.assigns[:subject] do
```
Comment (bmanifold) [inline elixir/apps/web/lib/web/auth.ex:217]: For my own edification, is this change a best practice or is there an actual difference in the way the "subject" is being accessed here?

File: elixir/apps/web/lib/web/components/core_components.ex [hunk 16]
```diff
@@ -221,7 +190,7 @@ defmodule Web.CoreComponents do
     <.paginator
       page={5}
       total_pages={100}
-      collection_base_path={~p"/users"}/>
+      collection_base_path={~p"/actors"}/>
```
Comment (bmanifold) [inline elixir/apps/web/lib/web/components/core_components.ex:193]: Are all references to "users" now going to be "actors" from now on?  I ask because my current PR likely will need to be changed if so.

File: elixir/apps/web/lib/web/components/core_components.ex [hunk 17]
```diff
@@ -713,6 +547,94 @@ defmodule Web.CoreComponents do
...
+          navigate={~p"/#{@schema.account_id}/actors/#{@schema.created_by_identity.actor.id}"}
+        >
+          <%= assigns.schema.created_by_identity.actor.name %>
+        </.link>
+        """
+    end
+  end
+
+  @doc """
+  Helps to pluralize a word based on a an cardinal number.
```
Comment (bmanifold) [inline elixir/apps/web/lib/web/components/core_components.ex]:
```suggestion
  Helps to pluralize a word based on a cardinal number.
```

**Review by bmanifold — APPROVED** [review]
Nothing blocking.  LGTM.  Just had a few small questions.

File: elixir/apps/domain/lib/domain/actors/actor/changeset.ex (hunk 12 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/actors/actor/changeset.ex:13]: @bmanifold Nothing changed in practice, before we had a match (the function will raise an error if an invalid value is passed), not we will get a changeset error because it's an enum value. I removed it here because don't want to duplicate enum values across the codebase.

File: elixir/apps/domain/lib/domain/auth.ex (hunk 13 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth.ex:52]: Exclamation is a convention that means that function will raise, this function returns `:ok` and `:error` tuples so it should not have one.

File: elixir/apps/domain/lib/domain/config/definitions.ex (hunk 14 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/config/definitions.ex:449]: It doesn't matter much, I keep them together because they are basically the same (google extends oidc).

File: elixir/apps/web/lib/web/auth.ex (hunk 15 above)
Comment (AndrewDryga) [reply elixir/apps/web/lib/web/auth.ex:217]: If the subject is not assigned before the function will just return whatever is in `else` block, but now it will raise. I don't think it should be ever called when the subject is not assigned, so it will help to fight some accidental mistakes.

File: elixir/apps/web/lib/web/components/core_components.ex (hunk 16 above)
Comment (AndrewDryga) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: This is hard again because of naming, we can keep them as `users` but then our domain language doesn't match our front-end, but if we name them Actors everywhere it can be confusing for our end users 🤷‍♂️.

File: elixir/apps/domain/lib/domain/auth.ex (hunk 13 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/auth.ex:52]: Ah, ok, cool.  Very good to know.  I had interpreted the exclamation mark to mean "use with caution".

File: elixir/apps/domain/lib/domain/config/definitions.ex (hunk 14 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/config/definitions.ex:449]: Ah, ok, sounds good. 👍

File: elixir/apps/web/lib/web/auth.ex (hunk 15 above)
Comment (bmanifold) [reply elixir/apps/web/lib/web/auth.ex:217]: Ah, ok, interesting.

File: elixir/apps/web/lib/web/components/core_components.ex (hunk 16 above)
Comment (bmanifold) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: Yeah, definitely agree this is a tough naming problem.

Will service/machine accounts be listed along side human users?  Or will they have their own URL path?

File: elixir/apps/domain/lib/domain/actors/actor/changeset.ex (hunk 12 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/actors/actor/changeset.ex:13]: Gotcha, that makes sense.  Thanks for the info!

File: elixir/apps/web/lib/web/components/core_components.ex (hunk 16 above)
Comment (jamilbk) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: I was planning to have them in a separate area in the UI, so it probably makes sense to scope the routes then?

Comment (bmanifold) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: Well, the reason I was asking was that I had the thought that if everything is under one page, then having the URL have `actors` in it might not be a big deal, but if things are going to get split out with having a page of only human users then it may be better to have the URL say `users`? 🤷‍♂️

Comment (jamilbk) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: @bmanifold I was thinking the same thing. `users` and `service_accounts` may be the same thing in the backend, but from an Admin's perspective, they're really two different things with different security contexts and workflows.

Comment (jamilbk) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: Hmm, but this makes for more UIs for us to build. I suppose we could keep things simpler for now, and just have one `actors` UI section with controls for segregating `service_account` actors and `user` actors.

On that note, we may also want to update `device` to `client` like @AndrewDryga had mentioned; I agree that `Device` is not a great name for headless clients running via `service_accounts`.

Comment (jamilbk) [reply elixir/apps/web/lib/web/components/core_components.ex:193]: We could discuss more on Slack or a quick call
